Flag options

20001LastOrder opened the pull request:
* Use  `argparse` to support different flag options to slack sherpa
* Refactor existing flag option to be handled with `argparse` for more reliable results
* Move message parsing (Slack Specific) from the agent to the slack app
* Added some more flag options that will be used in the future
* More tests

oshoma commented on src/tests/unit_tests/test_config.py:
```diff
@@ -0,0 +1,32 @@
+import pytest
```
We should name the test file so that it corresponds to the file being tested.
So in this case I suggest naming it `src/tests/unit_tests/sherpa_ai/test_config.py`.

This will make it easier to set up an automated test runner later, which runs just the tests that are necessary after each file change.

20001LastOrder replied:
This test folder is specific to test files in`sherpa_ai`. Maybe a good idea would be grouping the files by module: `src/tests/unit_tests/config/test_task_config.py`

oshoma commented on src/tests/unit_tests/test_config.py:
```diff
@@ -0,0 +1,32 @@
+from sherpa_ai.config import AgentConfig
+
+
+def test_parse_args():
```
Try to name the tests to say exactly what they verify. In this case, I would do something like this:
```
test_parse_args -> test_all_parameters_parse_successfully
test_parse_args_partial -> test_no_gsite_parses_successfully
test_parse_args_noise -> test_invalid_input_raises_value_error
```

oshoma reviewed (approved):
These changes all look solid

oshoma commented on src/tests/unit_tests/config/test_task_config.py:
```diff
@@ -0,0 +1,65 @@
+        AgentConfig.from_input(input_str)
+
+
+def test_invalid_input_raises_value_error():
```
This test is incorrectly named. I think it should be something like `test_gsite_parses_successfully`.

But I'm not sure this test should be here at all. It looks like an integration test, testing the combination of AgentConfig.from_input() and GoogleSearch.config_gsite_query().

20001LastOrder replied:
I think we should change this test to only test GoogleSearch Tool and move it to the tests of tools.

oshoma replied:
I agree, moving this test to a different file and narrowing it to only test GoogleSearchTool is a good next step

20001LastOrder replied:
Cool, I moved it to [src/tests/unit_tests/actions/test_google_search.py](https://github.com/Aggregate-Intellect/sherpa/pull/200/files#diff-ac4c7f8cd7a22d0916dae4bc21bff4690c4590a268c118779ff8bf36018a0479)

oshoma commented on src/tests/unit_tests/config/test_task_config.py:
```diff
@@ -0,0 +1,65 @@
+from sherpa_ai.tools import ContextTool, SearchTool
+
+
+def test_all_parameters_parse_successfully():
```
Suggestion: add `--do_reflect` as well

oshoma commented on src/sherpa_ai/tools.py:
```diff
@@ -77,12 +78,21 @@ def _arun(self, query: str) -> str:
         "you cannot find the information using internal search."
     )
 
+    def _add_gsite(self, query: str) -> str:
```
suggested rename: `query_with_gsite`

oshoma commented on src/sherpa_ai/tools.py:
```diff
@@ -77,12 +78,21 @@ def _arun(self, query: str) -> str:
 
+    def _add_gsite(self, query: str) -> str:
+        # check if the gsite is none
+        if self.config.gsite:
```
a more concise way of writing this:

`return query + " " + self.config.gsite if self.config.gsite else query`

the comment on 88 is not needed

YujingYang666777 replied:
The problem is addressed! Thanks for your suggestion

oshoma commented on src/sherpa_ai/config/task_config.py:
```diff
@@ -0,0 +1,41 @@
+    def from_config(cls, configs: List[str]) -> "AgentConfig":
+        parser = ArgumentParser()
+
+        parser.add_argument("--verbose", action="store_true")
```
suggestion: add help messages for these arguments so users know what the args do

20001LastOrder replied:
I think the help messages can be helpful. However, these help messages will not be directly shown to the user unless they are running Sherpa from the commandline. Let's add the helper message first and then find a way to show the help menu in general.

oshoma commented on src/sherpa_ai/actions/google_search.py:
```diff
@@ -31,10 +33,19 @@ def __init__(
         self.n = n
+        self.config = config
 
         self.search_tool = SearchTool()
```
See suggestions for `src/sherpa_ai/tools.py`

oshoma commented on src/apps/slackapp/tests/test_get_response.py:
```diff
@@ -0,0 +1,55 @@
+from sherpa_ai.verbose_loggers import DummyVerboseLogger
+
+
+def test_get_date():
```
`test_get_response_contains_todays_date`

oshoma commented on src/apps/slackapp/tests/test_get_response.py:
```diff
@@ -0,0 +1,55 @@
+    assert date in response, "Today's date not found in response"
+
+
+def test_get_question():
```
`test_response_contains_correct_info`

oshoma commented on src/apps/slackapp/tests/test_get_response.py:
```diff
@@ -0,0 +1,55 @@
+def test_get_question():
+    question = "What is AutoGPT and how does it compare with MetaGPT"
+
+    if cfg.SERPER_API_KEY is None:
```
I see this test and the one above are making actual calls to 3rd party APIs. We should not do this in tests, because we may end up unknowingly hammering the 3rd party API many times.

If we're trying to test our own logic, over and above what the 3rd party API does, we should mock/stub the 3rd party API calls/responses. This will make our tests deterministic, and avoids making network calls in tests.

If we want to know whether the 3rd party API does what it claims to do, that's a different problem. We expect the 3rd party to test their own stuff.

20001LastOrder replied:
I'll add the `real` mark to these tests so we can skip them if necessary.

oshoma replied:
OK. @20001LastOrder can you rename `real` to something more explicit, e.g. `external-api-caller`?

oshoma commented on src/sherpa_ai/task_agent.py:
```diff
@@ -152,13 +154,19 @@ def run(self, task: str) -> str:
-                formatted_logger_step = show_commands_only(logger_step)
-            except Exception as e:
-                logger.error(e)
+            # Serial Verbose Feature
```
conditionals like this are a little easier to read if you structure them with the positive case first, e.g.

```
if self.agent_config.verbose
    formatted_logger_step = logger_step
else
    try ...
```

oshoma commented on src/sherpa_ai/task_agent.py:
```diff
@@ -152,13 +154,19 @@ def run(self, task: str) -> str:
+            if not self.agent_config.verbose:
+                try:
+                    formatted_logger_step = show_commands_only(logger_step)
+                except Exception as e:
```
It's usually not a good idea to silently swallow any type of exception like this. Instead, let it blow up, figure out what specific exception types will be generated, and then add code again to handle each type of exception appropriately.

20001LastOrder replied:
I think in this case we are actually catching KeyError from the function

oshoma commented on src/apps/slackapp/slackapp/bolt_app.py:
```diff
@@ -46,22 +48,32 @@ def hello_command(ack, body):
+        message_cls = AIMessage if message["user"] == self.ai_id else HumanMessage
+        # replace the at in the message with the name of the bot
+        text = message["text"].replace(f"@{self.ai_id}", f"@{self.ai_name}")
+        # added by JF
```
remove this comment. `git blame` gives us file and line history.

oshoma commented on src/apps/slackapp/slackapp/bolt_app.py:
```diff
@@ -161,13 +149,21 @@ def event_test(client, say, event):
+            user_id,
+            team_id,
+            verbose_logger=SlackVerboseLogger(say, thread_ts),
+            bot_dict=bot,
```
can we give `bot_dict` a more descriptive name?  e.g. `bot_info`.

oshoma commented on src/apps/slackapp/slackapp/bolt_app.py:
```diff
@@ -46,22 +48,32 @@ def hello_command(ack, body):
-def contains_verbose(query: str) -> bool:
-    """looks for -verbose in the question and returns True or False"""
-    return "-verbose" in query.lower()
+def process_chat_history(messages: List[dict]) -> List[BaseMessage]:
```
Can we use a more descriptive function name here? E.g. `cleaned_and_classified_messages`

oshoma commented on src/apps/slackapp/slackapp/bolt_app.py:
```diff
@@ -46,22 +48,32 @@ def hello_command(ack, body):
+    return results
 
 
 def get_response(
```
This function is quite long. Think about breaking into smaller chunks and/or adding a docstring

oshoma reviewed (changes requested):
Hi folks, I did a thorough walkthrough. In general I like how this PR cleans things up. I made some suggestions for readability and maintainability.

oshoma reviewed (changes requested):
I think we are almost done here. My remaining unresolved comments are about the tests.

oshoma commented:
@20001LastOrder @YujingYang666777  I see there are also conflicts with the main branch. Please merge main into this branch to resolve conflicts.

oshoma commented on src/sherpa_ai/actions/google_search.py:
```diff
@@ -31,10 +33,18 @@ def __init__(
+        return query
+
     def execute(self, query) -> str:
+        query = self.config_gsite_query(query)
```
This code adds `site:` to the query by calling `config_gsite_query`.
Then SearchTool._run adds `site:` to the query by calling `query_with_gsqite`.  
(Assuming they are both using the same config.)
We should just add `site` to the query once, either here in `GoogleSearch` or in `SearchTool`.

If gsite is intended to be specific to google search then do the work here in `GoogleSearch`. 

Either way I suggest calling the function `query_with_gsite` or something like that.

20001LastOrder replied:
@YujingYang666777

20001LastOrder replied:
OK, I had a look at the implementation. It seems that this feature is meant to be added to the SearchTool, so the one in the GoogleSearch action is removed. And I moved the test accordingly.

oshoma reviewed (changes requested):
I added a few change requests.

Also, please ensure `make test` runs clean. When I run it I get 15 failed, 27 passed, 1 skipped, 7 warnings. Maybe I'm doing something wrong?

20001LastOrder commented:
> I added a few change requests.
> 
> Also, please ensure `make test` runs clean. When I run it I get 15 failed, 27 passed, 1 skipped, 7 warnings. Maybe I'm doing something wrong?

I was able to get the same failures. I fixed some of the failed tests due to the change introduced to the search tool. However, it seems other tests also fail on the main branch. We should tackle these failure cases in a separate pr.

oshoma commented:
I reduced test failures somewhat. We now have 7 failed, 33 passed, 1 skipped, 7 warnings in 678.56s (0:11:18) .
